### ISSUE TITLE
Check if string is encrypted

### ISSUE THREAD

**Issue by Firesphere** [defuse/php-encryption#284]
Is there a way to take any arbitrary hex string and check if it is encrypted by e.g. checking the header?
I may overlook it, but I can't find it.

**Comment by defuse** [comment]
Right now the only option is to try to decrypt it (using the correct key) and see if it works.

What's your use case? I'd be nervous about adding an API that lets you check headers, since really all a protocol using this API should care about is "Is this a valid ciphertext *produced by this key*?" There's no possible way to check "Is this a valid ciphertext produced by *some* key?" It's possible to check "Does this string look kinda sorta like a ciphertext, without any guarantees it actually is?" but I'm not sure what use that would have.

**Comment by Firesphere** [comment]
I would think, the header-bit, would be a give away?
My usecase, is where I fetch data from the database, that is encrypted.
After reading the the actual value, I get the decrypted value in my application, but for the duration of the request, this decrypted value is cached on the object itself.
Hence, next time I request the data, it tries to decrypt, but fails. For now, I have a workaround to resolve this by indeed having a try/catch method.
I am looking for a way to see if the string is an encrypted value, without having to go through the entire decryption.
The same issue occurs if I have an unencrypted field I want to encrypt in a later stage. I want the originals to not break the decryption, while still being accessible.

I understand why you are nervous about it and it's not a big deal. I was just asking if there was a better way, because right now, if it fails decryption, the original DB-value is returned, which is also not entirely the right way to go, because it could cause possible leaking of encryption.

I guess a better workaround would be, to mark the object as being decrypted. But that poses another possible issue around undecrypted values being leaked.

Long story short: Encryption is hard!

**Comment by paragonie-scott** [comment]
Adding an API to the library probably isn't necessary. [The header values are fixed](https://github.com/defuse/php-encryption/blob/master/src/Core.php#L12) for a given version of the library, after all.

```php
$header = Core::ourSubstr($ciphertext, 0, 4);
// The message should be hex-encoded
if (hash_equals($header, "def5")) {
    try {
        $plaintext = Crypto::decrypt($ciphertext, $key);
    } catch (Throwable $ex) {
        // Not encrypted
    }
} else {
    // Not encrypted
}
```

**Comment by Firesphere** [comment]
That's actually a good one. Thanks!

**Comment by defuse** [comment]
Note that the existence, and behavior, of those headers isn't guaranteed by the API documentation, so using them is equivalent to using an undocumented feature, and the behavior might change in future (even 0.0.1 minor) releases.

**Comment by defuse** [comment]
@Firesphere: Yes you do need a separate flag to indicate whether it's a ciphertext or not. What if the plaintext is a ciphertext itself, or starts with "def5"? Then your app will get confused and think the plaintext is a ciphertext. You need a separate flag for it to be reliable.

**Comment by paragonie-scott** [comment]
> Then your app will get confused and think the plaintext is a ciphertext. You need a separate flag for it to be reliable.

Yeah, that's what I was getting at with, "That being said, you should know this without having to test it at run-time."

**Comment by defuse** [comment]
Another security reason is: If an attacker tampers with a ciphertext, the MAC check will fail, and decryption will fail, and your app will think it's a plaintext, so it's trivial to mount forgery attacks against your app.
